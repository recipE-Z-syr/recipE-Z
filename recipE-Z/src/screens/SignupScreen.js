import 'react-native-gesture-handler';
import * as React from 'react';
import { Platform, SafeAreaView, Button, StyleSheet, Text, View, AppRegistry, Image, TextInput, Alert, ScrollView, Keyboard, TouchableOpacity, Component } from 'react-native';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import {Stitch, UserPasswordAuthProviderClient, UserPasswordCredential} from 'mongodb-stitch-react-native-sdk'
import defaultStyles from './stylesheet';


class SignupScreen extends React.Component {
  constructor(props) {
    super(props);

    this.client = Stitch.defaultAppClient;

    this.state = { fullName: '', email: '', password: '', confirmPassword: false}

    this.handleName = this.handleName.bind(this);
    this.handleEmail = this.handleEmail.bind(this);
    this.handlePassword = this.handlePassword.bind(this);
    this.handleConfirmPassword = this.handleConfirmPassword.bind(this);
    this.create_account = this.create_account.bind(this);
  }

  handleName(text) {
    this.setState({ fullName: text });
  }

  handleEmail(text) {
    this.setState({ email: text });
  }

  handlePassword(text) {
    this.setState({ password: text });
  }

  handleConfirmPassword(text) {
    if(text != this.state.password)
      this.setState({ confirmPassword: false });

    else 
      this.setState({ confirmPassword: true });
  }

  async create_account()
  {
    var status = "";

    var re = /^(?:[a-z0-9!#$%&amp;'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&amp;'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])$/;

    const email = this.state.email;
    const password = this.state.password;
    const confirmPassword = this.state.confirmPassword;

    if (this.state.fullName === '')
      alert("Cannot leave name blank!");

    else if (email === '')
      alert("Cannot leave email blank!");

    else if (!re.test(email))
      alert("Invalid email!")

    else if (password === '')
      alert("Cannot leave password blank!");

    else if (!confirmPassword)
      alert("Passwords do not match!");
    
    else if (password.length < 6 || password.length > 128)
      alert("Password must be between 6 and 128 characters!");
    else
    {
      const client = this.client;

      // const db = client.getServiceClient(RemoteMongoClient.factory, 'mongodb-atlas').db('users');

      const app = client

      const emailPasswordClient = app.auth.getProviderClient(UserPasswordAuthProviderClient.factory);

      await emailPasswordClient.registerWithEmail(email, password)
        .then(() => console.log("Successfully sent account confirmation email!"))
        .catch(err => {
                        console.error("Error registering new user:");
                        status = "register_error";
                        var message = err.message.charAt(0).toUpperCase() + err.message.substring(1) + '.';
                        if (message === "Name already in use.")
                          alert("Account with this email already exists.");
                        else
                          alert(message);
                      });

      if (status != "register_error")
      {
        const credential = new UserPasswordCredential(email, password)
        await app.auth.loginWithCredential(credential)
          // Returns a promise that resolves to the authenticated user
          .then(authedUser => console.log("Successfully logged in with id: ${authedUser.id}"))
          .catch(err => {
                          console.error("Login failed with error:", err);
                          status = "login_error";
                          alert(err.message.charAt(0).toUpperCase() + err.message.substring(1) + '.');
                        })
      }

      if (status === "")
        status = "success";

    }
    return status;
  }


  render(){
    const { navigation } = this.props;
    return (
      <View style={[defaultStyles.container, {backgroundColor:'#ffcece'}]}>
      <View style={[defaultStyles.whiteContainer, {paddingTop: 0, paddingBottom: 0}]}/>
        <Image
        style={{width: 90, height: 87, marginBottom: 10}}
        source={require('../img/logo.png')}
        />
        <View style={{alignItems: 'center'}}>
        <Text style={defaultStyles.h1}>
        Create an Account
        </Text>
        <TouchableOpacity
          onPress={() => navigation.navigate('Log-In')}>
          <Text style={defaultStyles.textLink}>Already have an account? Click here to login</Text>
        </TouchableOpacity>
        </View>
        <View style={{marginTop: 15, marginBottom: 10}}>
          <TextInput
          placeholder = "Full Name"
          style={defaultStyles.textInput}
          onBlur = {Keyboard.dismiss}
          value = {this.state.fullName}
          onChangeText = {this.handleName}
          />
          <TextInput
          placeholder = "Email"
          style={defaultStyles.textInput}
          onBlur = {Keyboard.dismiss}
          value = {this.state.email}
          onChangeText = {this.handleEmail}
          />
          <TextInput
          placeholder = "Password"
          style={defaultStyles.textInput}
          onBlur = {Keyboard.dismiss}
          value = {this.state.password}
          onChangeText = {this.handlePassword}
          />
          <TextInput
          placeholder = "Confirm Password"
          style={defaultStyles.textInput}
          onBlur = {Keyboard.dismiss}
          onChangeText = {this.handleConfirmPassword}
          />
          <Text style={{color: '#ed4848', fontSize: 12, textAlign: 'center', display: this.state.confirmPassword ? 'none' : 'flex'}}>
          The passwords you entered do not match.
          </Text>
        </View>
        <TouchableOpacity style={{alignItems: 'center', marginTop: 5, marginBottom: 20}}>
          <Text style={{color: '#ed4848', fontSize: 10}}>
          By creating an account you agree to our
          </Text>
          <Text style={{color: '#ed4848', fontSize: 10}}>
          Terms of Service and Privacy Policy
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[defaultStyles.redButton]}
          onPress={async () =>  {
                            var status = await this.create_account();
                            if(status === "success") 
                              navigation.navigate('Search');
                          }
                  }
          underlayColor='#ed4848'>
          <Text style={defaultStyles.redButtonText}>SIGN UP</Text>
        </TouchableOpacity>

      </View>
    );

  }
}


export default SignupScreen;

// const styles = StyleSheet.create({
//   loginButton: {
//     backgroundColor: '#ed4848',
//     marginRight: 40,
//     marginLeft: 40,
//     marginTop: 30,
//     paddingTop: 10,
//     paddingBottom: 20,
//     paddingLeft: 100,
//     paddingRight: 100,
//     borderRadius: 5,
//     borderWidth: 1,
//     borderColor: '#ed4848'
//   },
//   startText:{
//     color: 'white',
//     textAlign: 'center',
//     paddingLeft: 10,
//     paddingRight: 10,
//     paddingTop: 10,
//     fontWeight: 'bold'
//   },
//   accountLogin: {
//     color: '#ed4848',
//     textAlign: 'center',
//     fontSize: 11,
//     top: 60
//   }
// })
