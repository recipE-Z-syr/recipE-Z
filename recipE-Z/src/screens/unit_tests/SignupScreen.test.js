//SignupScreen.test.js

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, ShallowWrapper } from "enzyme";

import SignupScreen from '../SignupScreen';

describe('<SignupScreen />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<SignupScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should change state if name is entered', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handleName('James Smith');
    expect(instanceOf.state.fullName).toEqual('James Smith');
  });
  it('should change state if email is entered', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handleEmail('abc@gmail.com');
    expect(instanceOf.state.email).toEqual('abc@gmail.com');
  });
  it('should change state if password is entered', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handlePassword('password123');
    expect(instanceOf.state.password).toEqual('password123');
  });
  it('should change confirmPassword to true if passwords match', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handlePassword('password123');
    instanceOf.handleConfirmPassword('password123');
    expect(instanceOf.state.confirmPassword).toEqual(true);
  });
  it('should change confirmPassword to false if passwords have different cases', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handlePassword('password123');
    instanceOf.handleConfirmPassword('PASSWORD123');
    expect(instanceOf.state.confirmPassword).toEqual(false);
  });
  it('should change confirmPassword to false if passwords have different values', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handlePassword('password123');
    instanceOf.handleConfirmPassword('password456');
    expect(instanceOf.state.confirmPassword).toEqual(false);
  });
  it('should change confirmPassword to false if passwords have different values', () => {
    const instanceOf = renderer.create(<SignupScreen />).getInstance();
    instanceOf.handlePassword('password123');
    instanceOf.handleConfirmPassword('abc');
    expect(instanceOf.state.confirmPassword).toEqual(false);
    instanceOf.handlePassword('abc');
    instanceOf.handleConfirmPassword('password123');
    expect(instanceOf.state.confirmPassword).toEqual(false);
  });
});

const createTestProps = () => ({
  navigation: {
    navigate: jest.fn()
  }
});

test('Already have account navigates to Sign Up', () => {
  // const spy = jest.spyOn(navigation, 'navigate')
  let wrapper: ShallowWrapper;
  let props: any
  props = createTestProps();
  wrapper = shallow(<SignupScreen {...props} />);  

  const startButton = wrapper.find('TouchableOpacity').at(0)
  startButton.props().onPress()

  expect(props.navigation.navigate).toBeCalledWith('Log-In')
})
test('Sign Up navigates to Search Screen', () => {
  // const spy = jest.spyOn(navigation, 'navigate')
  let wrapper: ShallowWrapper;
  let props: any
  props = createTestProps();
  wrapper = shallow(<SignupScreen {...props} />);  

  const startButton = wrapper.find('TouchableOpacity').at(2)
  startButton.props().onPress()

  expect(props.navigation.navigate).toBeCalledWith('Search')
})