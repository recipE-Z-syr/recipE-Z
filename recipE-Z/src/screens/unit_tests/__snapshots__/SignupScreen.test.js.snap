// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupScreen /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "fontFamily": "nunito-regular",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#ffcece",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "height": 561,
          "justifyContent": "space-around",
          "paddingBottom": 70,
          "paddingTop": 70,
          "position": "absolute",
          "width": 334,
        },
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        },
      ]
    }
  />
  <Image
    source={1}
    style={
      Object {
        "height": 87,
        "marginBottom": 10,
        "width": 90,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#ed4848",
          "fontFamily": "nunito-extrabold",
          "fontSize": 20,
        }
      }
    >
      Create an Account
    </Text>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ed4848",
            "fontFamily": "nunito-regular",
            "fontSize": 11,
            "textAlign": "center",
          }
        }
      >
        Already have an account? Click here to login
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 10,
        "marginTop": 15,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Full Name"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "lightgrey",
          "borderWidth": 1,
          "fontFamily": "nunito-regular",
          "height": 50,
          "marginBottom": 5,
          "marginTop": 5,
          "opacity": 0.7,
          "paddingLeft": 15,
          "width": 280,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Email"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "lightgrey",
          "borderWidth": 1,
          "fontFamily": "nunito-regular",
          "height": 50,
          "marginBottom": 5,
          "marginTop": 5,
          "opacity": 0.7,
          "paddingLeft": 15,
          "width": 280,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "lightgrey",
          "borderWidth": 1,
          "fontFamily": "nunito-regular",
          "height": 50,
          "marginBottom": 5,
          "marginTop": 5,
          "opacity": 0.7,
          "paddingLeft": 15,
          "width": 280,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      placeholder="Confirm Password"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "lightgrey",
          "borderWidth": 1,
          "fontFamily": "nunito-regular",
          "height": 50,
          "marginBottom": 5,
          "marginTop": 5,
          "opacity": 0.7,
          "paddingLeft": 15,
          "width": 280,
        }
      }
      underlineColorAndroid="transparent"
    />
    <Text
      style={
        Object {
          "color": "#ed4848",
          "display": "none",
          "fontSize": 12,
          "textAlign": "center",
        }
      }
    >
      The passwords you entered do not match.
    </Text>
  </View>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "marginBottom": 20,
        "marginTop": 5,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#ed4848",
          "fontSize": 10,
        }
      }
    >
      By creating an account you agree to our
    </Text>
    <Text
      style={
        Object {
          "color": "#ed4848",
          "fontSize": 10,
        }
      }
    >
      Terms of Service and Privacy Policy
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#ed4848",
        "borderColor": "#ed4848",
        "borderRadius": 5,
        "borderWidth": 10,
        "justifyContent": "center",
        "opacity": 1,
        "paddingBottom": 10,
        "paddingTop": 10,
        "width": 200,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "nunito-bold",
          "paddingLeft": 10,
          "paddingRight": 10,
          "textAlign": "center",
        }
      }
    >
      SIGN UP
    </Text>
  </View>
</View>
`;
